crates/dbsp/src/operator/dynamic/distinct.rs

```diff
@@ -179,6 +241,46 @@ where
             .clone()
     }
 
+    // TODO: The hash_distinct operator internally does `map_index().distinct().map_index()`.
+    // The second `map_index` can be folded into `distinct`, but I think this can wait until we have
+    // empirical proof that this operator helps performance.
+    pub fn dyn_hash_distinct(&self, factories: &HashDistinctFactories<Z, C::Time>) -> Stream<C, Z>
+    where
+        Z: IndexedZSet + Send + DynFilterMap,
+    {
+        let circuit = self.circuit();
+        circuit.region("hash_distinct", || {
+            let stream = self.dyn_shard(&factories.input_factories);
+
+            circuit
+                .cache_get_or_insert_with(DistinctIncrementalId::new(stream.stream_id()), || {
+                    let by_hash: Stream<C, OrdIndexedZSet<DynData, DynPair<Z::Key, Z::Val>>> =
+                        stream
+                            .dyn_map_index(
+                                &factories.distinct_factories.input_factories,
```

Comment:
so this applies to an IndexedZSet already?

Reply:
A Z-set is a special case of an indexed Z-set, so it applies to both.